### PR TITLE
[#11198] Improve error handling for search service

### LINKED ISSUE #11198: Instructor searching for students: distinguish between 'no results' and 'internal error'
v8.*
image

When an instructor is searching (as above), if there was an internal error (e.g., the search service cannot be reached), the error message should say something like 'internal error' rather than 'no results' (which is the case now).

Otherwise it could mislead the user.

### PR DESCRIPTION
**Pull request by jianhandev**
Fixes #11198

Error message displayed when search service is not running.
Screenshot 2021-07-03 at 11 10 10 PM

Error message displayed when search service is active but no results returned.
no results

### CODE REVIEW

File: src/main/java/teammates/storage/search/SearchManager.java [hunk 1]
```diff
@@ -69,21 +72,32 @@ SolrQuery getBasicQuery(String queryString) {
         return query;
     }
 
-    QueryResponse performQuery(SolrQuery query) throws SearchNotImplementedException {
+    QueryResponse performQuery(SolrQuery query) throws SearchServiceException {
         if (client == null) {
-            throw new SearchNotImplementedException();
+            throw new SearchServiceException("Search service is not implemented.", HttpStatus.SC_NOT_IMPLEMENTED);
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/search/SearchManager.java:77]: Even though it's written like this before, let's rephrase this a bit to "~Search service~Full-text search is not ~implemented~ available".

File: src/main/java/teammates/storage/search/SearchManager.java [hunk 2]
```diff
@@ -69,21 +72,32 @@ SolrQuery getBasicQuery(String queryString) {
...
+            log.severe(String.format(ERROR_SEARCH_DOCUMENT, query.getQuery(), rootCause)
                     + TeammatesException.toStringWithStackTrace(e));
+            if (rootCause instanceof ConnectException) {
+                throw new SearchServiceException("Error connecting to search service.",
+                        e, HttpStatus.SC_BAD_GATEWAY);
+            } else if (rootCause instanceof SocketTimeoutException) {
+                throw new SearchServiceException("A timeout was reached while processing the request.",
+                        e, HttpStatus.SC_GATEWAY_TIMEOUT);
+            } else {
+                throw new SearchServiceException("An error has occurred.", e, HttpStatus.SC_BAD_GATEWAY);
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/search/SearchManager.java:95]: This is very vague when displayed on the front-end. At the very least, should mention "An error has occurred *while performing search*".

File: src/main/java/teammates/storage/search/SearchManager.java [hunk 3]
```diff
@@ -69,21 +72,32 @@ SolrQuery getBasicQuery(String queryString) {
...
 
         try {
             response = client.query(getCollectionName(), query);
         } catch (SolrServerException e) {
-            log.severe(String.format(ERROR_SEARCH_DOCUMENT, query.getQuery(), e.getRootCause())
+            Throwable rootCause = e.getRootCause();
+            log.severe(String.format(ERROR_SEARCH_DOCUMENT, query.getQuery(), rootCause)
                     + TeammatesException.toStringWithStackTrace(e));
+            if (rootCause instanceof ConnectException) {
+                throw new SearchServiceException("Error connecting to search service.",
```
Comment (wkurniawan07) [inline src/main/java/teammates/storage/search/SearchManager.java:89]: We may not need separate error message/branch here. Unlike 4XX errors where we can tailor the message to tell users what to do to rectify the error, for 5XX errors and particularly for this case it doesn't help whether they know it's an error caused by connection (this branch) or something else (later branch); all they know is (1) it is the system's fault and not theirs and (2) it is/should be a temporary error.

**Review by wkurniawan07 — CHANGES_REQUESTED** [review]

File: src/test/java/teammates/storage/search/StudentSearchTest.java [hunk 4]
```diff
@@ -10,7 +10,7 @@
 import teammates.common.datatransfer.attributes.CourseAttributes;
```
Comment (daongochieu2810) [inline src/test/java/teammates/storage/search/StudentSearchTest.java:10]: I think its better if we test that the exception messages are correctly returned

Comment (jianhandev) [reply src/test/java/teammates/storage/search/StudentSearchTest.java:10]: You're right, I'll add on to the tests in another PR which aims to improve testing for the new search service.

**Review by wkurniawan07 — APPROVED** [review]
LGTM

**Comment by wkurniawan07** [comment]
@jianhandev while this PR improves the error handling of searching, equally as important or even more important is to improve the error handling of indexing. Right now we kind of sweep the problem under the rug by just error logging.

**Comment by jianhandev** [comment]
> We also need to find a way to specify the timeout limit for the Solr client. If my gut feel is right currently connections will not time out, and that's not the best way to do it.

Alright, I'll try to figure this out.

**Pull request merged**